[WIP] Rotate SSH keys page (#7669)
circleci/circleci-docs: jekyll/_cci2/rotate-project-ssh-keys.adoc

[#github-projects]
== Github projects

Go to **Project Settings > SSH Keys** to view SSH keys set up for your project.

[#rotate-a-deploy-key-github]
=== Rotate a deploy key

. Take note of the current key information to rotate, including the fingerprint. You can also click the keyname to open the related GitHub page that should list the public key.
. Delete the deploy key by clicking the **X**.
. Add a new deploy key by clicking **Add Deploy Key**.
. Go to GitHub’s repository project settings to delete the matching public key. The github URL is typically `https://github.com/<org>/<project>/settings/keys`, or you may already have the page open if you clicked on the keyname in step 1. The keys are named `CircleCI`. Removing any key titled `CircleCI` created before the rotation is recommended. The new public SSH key will be automatically added once the old key is deleted.

[#rotate-a-user-key-github]
=== Rotate a user key

If you have set up user keys for your project, follow these steps:

. Take note of the current key information to rotate, including the fingerprint. You can also click the keyname to open the related Github page that should list the public key.

> . Delete the deploy key by clicking the **X**.

Reviewer: ```suggestion
. Delete the user key by clicking the **X**.
```

. Add a new user key by clicking **Add User Key**.

> . Go to GitHub’s User settings to delete the matching public key. The github URL is typically `https://github.com/settings/keys`or you may already have the page open if you clicked on the keyname in step 1. Removing any key titled `CircleCI` created before this rotation is recommended. The new public SSH key will be automatically added once the old key is deleted (page may need to be refreshed to view the key).

Reviewer: ```suggestion
. Go to GitHub’s User settings to delete the matching public key. The GitHub URL is typically `https://github.com/settings/keys`or you may already have the page open if you clicked on the keyname in step 1. Removing any key titled `CircleCI` created before this rotation is recommended. The new public SSH key will be automatically added once the old key is deleted (page may need to be refreshed to view the key).
```
^ Also just want to double check that the part needing to refresh the page before the new key can be viewed only applies to user keys and not other keys as well

Author: I think this note should be added to all of them actually

CAUTION: The user key name contains the project name, however, a user key may give access to more projects within the organization.

> NOTE: If using organization SSO, you can take note of which one is currently authorized. For the newly created key, if using organization SSO, and access is needed, you will need to reauthorize it.

Reviewer: >  which one is currently authorized

Not super clear which "one" this is referring to. User vs project key? I looked back at the Google doc and this would be my guess?

Reviewer 2: 'one' in that context is 'organization'. For example I am part of 4 orgs and my key is authorized only for two of them.

Author: I will try reworking this note so it's easier to read

> NOTE: If using organization SSO, you can take note of which one is currently authorized. For the newly created key, if using organization SSO, and access is needed, you will need to reauthorize it.

Author: ```suggestion
NOTE: If using organization SSO, take note of which org is currently authorized, and, if access is needed for the newly created key, you will need to reauthorize it.
```

[#rotate-an-additional-SSH-key-github]
=== Rotate an additional SSH key

If you are using additional SSH keys in your project, follow these steps:

> . Take note of the existing key to know which target system it is used for. This could be a VCS, a machine or other SSH based system.

Reviewer 2: To help with the last step, explicitly mention the hostname
```suggestion
. Take note of the existing key's Hostname to know which target system it is used for and the fingerprint for your records. This could be a VCS, a machine or other SSH based system.
```

> . Take note of the existing key's hostname to know which target system it is used for, and the fingerprint for your records. The target could be a VCS, a machine or other SSH based system.

Author: ```suggestion
. Take note of the existing key's hostname to know which target system it is used for, and the fingerprint for your records. The target could be a VCS, a machine, or another SSH based system.
```

> . Delete the deploy key by clicking the **X**.

Reviewer: ```suggestion
. Delete the SSH key by clicking the **X**.
```

. Remove the authorization to use that key (the public key) from the target system. Typically, the public key will be authorized in the target system, and the configuration varies depending on the target system.

> . Recreate the key and the authorization in the target system.

Reviewer: Are we supposed to include an additional step here for "Add SSH Key" within the CircleCI project?

Reviewer 2: yes! I realized that too. The 3 statements are missing the step to do it in cci UI. It is the same as explained for gitlab additional keys
**Add the private key by clicking Add SSH Key. Paste in the private key you created in the step above and click Add SSH Key again to finalize.**

Author: @sbeaulie Can they do this in the CCI web app straight after deleting the original additional SSH private key, before leaving the app to go an delete the public key from the target? 

And are the steps as follows:

. create a new SSH key using `ssh-keygen -t ed25519 -C your_email@example.com` or however they want to generate it
. click ADD SSH key
. As Hostname you can use the same value you had for the old key in step 1
. click Add SSH Key again to finalize

[#bitbucket-projects]
== Bitbucket projects

Go to **Project Settings > SSH Keys** to view SSH keys set up for your project.

[#rotate-a-deploy-key-bitbucket]
=== Rotate a deploy key

. Take note of the current key information to rotate.
. Delete the deploy key by clicking the **X**.
. Add a new deploy key by clicking **Add Deploy Key**.
. Go to Bitbucket’s project admin access-key settings to delete the matching public key. The Bitbucket URL is typically `https://bitbucket.org/<org>/<project>/admin/access-keys/`. The keys are named `CircleCI`. Since the fingerprint hash is not shown, it is recommended to remove any key titled `CircleCI` created before the rotation. The new public SSH key will be automatically added once the old key is deleted (page may need to be refreshed to view the key).

[#rotate-a-user-key-bitbucket]
=== Rotate a user key

. Take note of the current key information to rotate.

> . Delete the deploy key by clicking the **X**.

Reviewer: ```suggestion
. Delete the user key by clicking the **X**.
```

Reviewer 2: yes I just realized that too, all three should be followed by
Add the private key by clicking Add SSH Key. It is recommended you name the key [gitlab.com](http://gitlab.com/). Paste in the private key you created in the step above and click Add SSH Key again to finalize.

. Add a new user key following the xref:bitbucket-integration/#create-a-bitbucket-user-key[Create a Bitbucket user key] instructions.
. Go to Bitbucket’s user account settings to delete the matching public key. The Bitbucket URL is typically `https://bitbucket.org/account/settings/ssh-keys/`. The user names the keys, therefore, CircleCI does not know if the key name contains the string `CircleCI`. It is recommended to remove any key created before the rotation.

[#rotate-an-additional-SSH-key-bitbucket]
=== Rotate an additional SSH key

> . Take note of the existing key to know which target system it is used for. This could be a version control system, a machine or other SSH based system.

Reviewer 2: ```suggestion
. Take note of the existing key's Hostname to know which target system it is used for and the fingerprint for your records. This could be a version control system, a machine or other SSH based system.
```

> . Delete the deploy key by clicking the **X**.

Reviewer: ```suggestion
. Delete the SSH key by clicking the **X**.
```

. Remove the authorization to use that key (the public key) from the target system. Typically, the public key will be authorized in the target system, and the configuration varies depending on the target systems.

> . Recreate the key and the authorization in the target system.

Reviewer: ^ Same question as above regarding **Add SSH Key**

Reviewer 2: yes I suggest
**Add the private key by clicking Add SSH Key. Paste in the private key you created in the step above and click Add SSH Key again to finalize.**

[#gitlab-projects]
== GitLab projects

Go to **Project Settings > SSH Keys** to view SSH keys set up for your project. The GitLab integration with CircleCI does not use deploy keys or user keys.

[#rotate-an-SSH-key-gitlab]
=== Rotate an SSH key

. Take note of the existing key and fingerprint to be able to match the target system it is used for, typically `gitlab.com`.

> . Delete the deploy key by clicking the **X**.

Reviewer: ```suggestion
. Delete the SSH key by clicking the **X**.
```

. The key will need to be recreated manually, for example, using OpenSSH (`ssh-keygen -t ed25519 -C "your_email@example.com"`). Instructions can be found on link:https://docs.gitlab.com/ee/user/ssh.html[GitLab's documentation].
. Add the private key by clicking **Add SSH Key**. It is recommended you name the key `gitlab.com`. Paste in the private key you created in the step above and click **Add SSH Key** again to finalize.
. Go to GitLab’s project settings for the repository and expand deploy keys section. The GitLab URL is typically `https://gitlab.com/<org>/<project>/-/settings/repository`. The keys are named `circleci-pipeline-triggers` and you can match the fingerprint. It is recommended to remove any key created before the rotation.
. Add the new key. For consistency, you can name the key in the title section `circleci-pipeline-triggers`. Paste the public key and click **Add key** to finalize.

[#rotate-an-additional-SSH-key-gitlab]
=== Rotate an additional SSH key

In some scenarios, you might also be using additional SSH keys that are not tied to GitLab.

> . Take note of the existing key to know which target system it is used for. This could be a version control system, a machine or other SSH based system.

Reviewer 2: ```suggestion
. Take note of the existing key's Hostname to know which target system it is used for and the fingerprint for your records. This could be a version control system, a machine or other SSH based system.
```

> . Delete the deploy key by clicking the **X**.

Reviewer: ```suggestion
. Delete the SSH key by clicking the **X**.
```

. Remove the authorization to use that key (the public key) from the target system. Typically, the public key will be authorized in the target system, and the configuration varies depending on the target systems.

> . Recreate the key and the authorization in the target system.

Reviewer: ^ Same question as above regarding **Add SSH Key**. The preceding section does mention it this time though

Reviewer 2: yes. I suggest
**Add the private key by clicking Add SSH Key. Paste in the private key you created in the step above and click Add SSH Key again to finalize.**

[#github-projects]
== Github projects

Go to **Project Settings > SSH Keys** to view SSH keys set up for your project.

[#rotate-a-deploy-key-github]
=== Rotate a deploy key

. Take note of the current key information to rotate, including the fingerprint. You can also click the keyname to open the related GitHub page that should list the public key.
. Delete the deploy key by clicking the **X**.
. Add a new deploy key by clicking **Add Deploy Key**.
. Go to GitHub’s repository project settings to delete the matching public key. The GitHub URL is typically `https://github.com/<org>/<project>/settings/keys`, or you may already have the page open if you clicked on the keyname in step 1. The keys are named `CircleCI`. Removing any key titled `CircleCI` created before the rotation is recommended. The new public SSH key will be automatically added once the old key is deleted.

[#rotate-a-user-key-github]
=== Rotate a user key

If you have set up user keys for your project, follow these steps:

> . Take note of the current key information to rotate, including the fingerprint. You can also click the keyname to open the related Github page that should list the public key.

Reviewer: ```suggestion
. Take note of the current key information to rotate, including the fingerprint. You can also click the keyname to open the related GitHub page that should list the public key.
```
